### PR TITLE
Allow specific testcontainers on product test.

### PR DESCRIPTION
**Pull request by s2lomon**
This pr allows usage of different subclasses of GenericContainer, rather than limiting us to only single implementation, that does not allow to use existing predefined testcontainers.

### CODE REVIEW

**Review by kokosing — COMMENTED** [review]
A lot of tests are failing.

Let's discuss this offline first.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DelegateContainers.java [hunk 1]
```diff
@@ -57,6 +61,59 @@ protected void containerIsStopped(InspectContainerResponse containerInfo)
         };
     }
 
+    /**
+     * This factory returns FixedPortPostgreSQLContainer (instead of PosgreSQLContainer) to not force @SuppressWarnings("unchecked") on its every usage.
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DelegateContainers.java:65]: typo in `PosgreSQLContainer`

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 2]
```diff
@@ -74,18 +93,32 @@
...
     private List<String> logPaths = new ArrayList<>();
-    private Optional<EnvironmentListener> listener = Optional.empty();
 
     public DockerContainer(String dockerImageName, String logicalName)
     {
-        super(dockerImageName);
+        this(logicalName, fixedHostPort(dockerImageName), container -> {});
+    }
+
+    public <T extends GenericContainer<T>> DockerContainer(String logicalName, DelegateContainerFactory<T> containerFactory, Consumer<T> containerClosure)
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:108]: what is the benefit of splitting `containerFactory` and `containerClousure` given fact that one is called directly after the other.
Cant you put the logic of closure into factory itself.

Also name `containerClousure` is very non-descriptive

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 3]
```diff
@@ -74,18 +93,32 @@
...
+    {
         this.logicalName = requireNonNull(logicalName, "logicalName is null");
+        requireNonNull(containerFactory, "containerFactory is null");
+        requireNonNull(containerClosure, "containerFactory is null");
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:112]: should be `containerClosure` in error message

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java [hunk 4]
```diff
@@ -41,6 +41,8 @@
 {
     Logger log = Logger.get(EnvironmentListener.class);
 
+    EnvironmentListener NO_OP = new EnvironmentListener() {};
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java:44]: too acronymish IMO. `EMPTY`?

**Review by losipiuk — COMMENTED** [review]
Thanks. Looks reasonable.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java (hunk 4 above)
Comment (findepi) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java:44]: EMPTY is more suitable for values, not action objects. We use NOOP in some places

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java (hunk 3 above)
Comment (findepi) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:112]: https://gist.github.com/findepi/b597d2bc91e5a30c2a44fd65fe252f0b

**Comment by kokosing** [comment]
Do we still need this?  Maybe close this, this already has bunch of conflicts.

**Comment by kokosing** [comment]
-

**Pull request closed without merging**